pkg/agent/storage/https.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2020 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	HEADER_SUFFIX = "-headers"
+)
+
+type HTTPSClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type HTTPSProvider struct {
+	Client *HTTPSClient
+}
+
+func (m *HTTPSProvider) DownloadModel(modelDir string, modelName string, storageUri string) error {
+	log.Info("Download model ", "modelName", modelName, "storageUri", storageUri, "modelDir", modelDir)
+	uri, err := url.Parse(storageUri)
+	if err != nil {
+		return fmt.Errorf("unable to parse storage uri: %v", err)
+	}
+	HTTPSDownloader := &HTTPSDownloader{
+		StorageUri: storageUri,
+		ModelDir:   modelDir,
+		ModelName:  modelName,
+		Uri:        uri,
+	}
+	if err := HTTPSDownloader.Download(*m.Client); err != nil {
+		return err
+	}
+	return nil
+}
+
+type HTTPSDownloader struct {
+	StorageUri string
+	ModelDir   string
+	ModelName  string
+	Uri        *url.URL
+}
+
+func (h *HTTPSDownloader) Download(client HTTPSClient) error {
+	// Create request
+	req, err := http.NewRequest("GET", h.StorageUri, nil)
+	if err != nil {
+		return err
+	}
+
+	headers, err := h.extractHeaders()
+	if err != nil {
+		return err
+	}
+	for key, element := range headers {
+		req.Header.Add(key, element)
+	}
+
+	// Query request
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make a request: %v", err)
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("URI: %s returned a %d response code", h.StorageUri, resp.StatusCode)
+	}
+
+	// Write content into file(s)
+	contentType := resp.Header.Get("Content-type")
+	fileDirectory := filepath.Join(h.ModelDir, h.ModelName)
+
+	if strings.Contains(contentType, "application/zip") {
+		if err := extractZipFiles(resp.Body, fileDirectory); err != nil {
+			return err
+		}
+	} else if strings.Contains(contentType, "application/x-tar") || strings.Contains(contentType, "application/gzip") {
+		if err := extractTarFiles(resp.Body, fileDirectory); err != nil {
+			return err
+		}
+	} else {
+		paths := strings.Split(h.Uri.Path, "/")
```

Comment:
Do we validate if the url path always include a file?

Reply:
We do not validate if the url path always include a file. Should we? It seems difficult to do so as the file can technically be in any format in this case.